### ISSUE TITLE
Add a pattern customization for filename

### ISSUE THREAD

**Issue by leopardlab** [mozilla-services/screenshots#4444]
When I download a screenshot I want it to be named and saved based on my own pattern so that I won't need to rename and sort constantly my screenshots.

Given the site url="https://example.com/foo/bar" having the site title="Bar - Example Domain" 
          And the screenshot pattern: "myscreenshots/{domain}/{pathname}/{date}{extension}" 
When I click download button, then I want a screenshot file to be saved in  the config("browser.download.dir") + "/myscreenshots/example.com/foo_bar/20180514.png" destination.

Needed patterns:
- {title}
  returns: "Bar - Example Domain"
- {domain}
  returns: "example.com"
- {path}
  returns: "/foo/bar"
- {pathname}
   returns: "/foo_bar"
- {extension}
   returns: ".png"
- {date}
   returns: "20180514"
- {datetime}
   returns: "20180514T102641"

**Comment by jscher2000** [comment]
This would be useful.

Question: where would an options UI be implemented for this kind of extension? You could place a button on the initial overlay, or in the post-capture bar, but then where would the user do the actual configuration? Popup doesn't seem ideal at that point.

**Comment by Viewfinder98** [comment]
I too find Firefox Screenshots tool very useful and cuts down on add-ons loaded.  As ideal as they are, I don't need as many customisations as OP but merely just a timestamp added to the filename to make it unique.

At present, the default filename only includes the date so when taking several screenshots of webpages which share the same page title, the filenames are always identical so I'm having to add a sequential number at the end or a time to it so I don't overwrite previous.

For example, when using Firefox Screenshots the default filename format is:

Screenshot-YYYY-M-DD Page Title.png

However if I use the Web Developer Screenshot tool, the filename format is:

Screen Shot 2018-08-13 at 13.39.33-fullpage.png

which includes timestamp including seconds so ensuring no duplicate files but it does not include the page title nor where to save.

Both screenshot tools have their own advantages and disadvantages, it would be great if the best of both can be combined under the same Mozilla developer.

**Comment by MarkRH** [comment]
Yeah.. needs to be a way to customize how the filename gets written..  

I'd rather have the Domain and/or Page Title in the beginning followed by "- Screenshot date/time".png.

Hard to look for stuff when everything starts with "Screenshot".  Making me tempted to go back to an alternate screenshot extension and disable the built-in one again.

**Comment by djh42** [comment]
`Screenshot-YYYY-M-DD Page Title.png`
Actually, it's worse than that. It's not YYYY-M-DD it's actually YYYY-M-D. A really braindead format if ever there was one.

**Comment by Kaos-Industries** [comment]
Same as others, I'd also love for screenshot names to be customisable. All I need is that the screenshot be saved with the page title at the beginning instead of the redundant "Screenshot" prefix and datestamp - I need to be able to sort my screenshots by name and that prefix doesn't help. I really hope someone is working on this.

**Comment by fabswt** [comment]
Would make sense to use the same filename format as in the *Responsive Design Mode*s screenshots, for consistency.

**Comment by djh42** [comment]
>    I need to be able to sort my screenshots by name and having prefixes before the more unique title doesn't help.

I don't understand that comment. An unchanged prefix in front of everything doesn't change the sort order at all. So what's your real objection? (Not that I object to your call for customisability)

**Comment by Kaos-Industries** [comment]
> 
> 
> > I need to be able to sort my screenshots by name and having prefixes before the more unique title doesn't help.
> 
> I don't understand that comment. An unchanged prefix in front of everything doesn't change the sort order at all. So what's your real objection? (Not that I object to your call for customisability)

Yes, I mistakenly assumed that the prefix changed the sort order - it doesn't, but it still is redundant and unsightly. especially for longer title names:

Screenshot_2019-04-17 Worksheet - Solving equations with unknowns on both sides.png
Screenshot_2019-04-17 Worksheet - Solving equations with brackets and unknowns on both sides.png
Screenshot_2018-12-06 Transferring large file off of MS-DOS 5 0 Machine (without removing HDD).png

An option to specify a pattern would allow me to get rid of such redundant information and leave only the information that matters to me in the filename for easier scanning and better aesthetics.

**Comment by fabswt** [comment]
The use of different formats for *Responsive Design* and regular screenshots *does* break the sort order though, compare:

Regular:

> Screenshot_2019-06-18 My Test Page.png

Responsive mode:

> Screen Shot 2019-06-17 at 18.13.50.png

**Comment by Kaos-Industries** [comment]
> 
> 
> The use of different formats for _Responsive Design_ and regular screenshots _does_ break the sort order though, compare:
> 
> Regular:
> 
> > Screenshot_2019-06-18 My Test Page.png
> 
> Responsive mode:
> 
> > Screen Shot 2019-06-17 at 18.13.50.png

That's a good catch, have never taken a screenshot in responsive mode so wouldn't have known about it.

**Comment by Janaue** [comment]
please use the locale to name the picture. calling it ''screenshot'' even with non english firefox is a lack of awareness.

**Comment by LC43** [comment]
for me, i would like to have the pattern to change when creating multiple screenshots. instead of <name>(1).png etc, would prefer <name>_1.png.

its a bit painful to type \(1 when in command line.

**Comment by sharuzzaman** [comment]
Any traction on this bug report? After 2 years nobody cares for this issue?

**Comment by MarkRH** [comment]
I just quit using it and use an extension instead that lets me better control the filename pattern. This is obviously not something on the to-do list.

**Comment by Kaos-Industries** [comment]
> Any traction on this bug report? After 2 years nobody cares for this issue?

Agreed, this is highly disappointing for a project as big as Firefox and for something that would be so simple to implement. I'm still getting by on renaming the screenshots manually because if rather not weigh down my installation by installing a plugin for something so small.

**Comment by gitressa** [comment]
It would be immediately helpful if time was added to the timestamp like this:
`Screenshot_yyyy-mm-dd_hh-mm-ss.png`

**Comment by amaury-hanser** [comment]
Any chance to see this feature ?

**Comment by Lurgainn** [comment]
> It would be immediately helpful if time was added to the timestamp like this:
> `Screenshot_yyyy-mm-dd_hh-mm-ss.png`

At least can we have this?
